### PR TITLE
[ENG-3120] Model-level logic for Approvals on SchemaResponses

### PR DESCRIPTION
**Pull request by jwalz**
## Purpose
This PR adds integration with the ApprovalsMachine for SchemaResponse objects as well as the creation of SchemaResponseActions following state machine triggers.

## Changes
* Add initial support to `SchemaResponses` to make use of the `ApprovalsMachine`
  * Add `reviews_state` field
  * Add `pending_approvers` relationship  
  * Define necessary callbacks and precondition checks
* Add state checks to `SchemaResponse.create_from_previous_response`, `schema_response.update_responses` and `schema_response.delete`
* Add SchemaResponeAction and SchemaResponseTriggers to support state machine triggers through the API
* So so so many tests
* Homogenize exception messages
* Make nullable-fields on SchemaResponse blank-able as well (required to pass validation outside of pytest)
* Add index on reviews_state and default ordering on "-created" to SchemaResponse

## QA Notes

Confirmed that existing tests for Sanctions/registrations pass

In process of adding tests for `SchemaResponses`

### CODE REVIEW

File: osf/models/schema_response.py [hunk 1]
```diff
@@ -54,6 +75,29 @@ def updated_response_keys(self):
...
+    def state(self, new_state):
+        self.reviews_state = new_state.db_name
+
+    @property
+    def is_moderated(self):
+        '''Determine if this SchemaResponseResponse belong to a moderated resource'''
+        return getattr(self.parent, 'is_moderated', False)
+
+    @property
+    def revisable(self):
```
Comment (jwalz) [inline osf/models/schema_response.py:93]: Make this a class attribute; property can be added if logic is needed

File: osf/models/schema_response.py [hunk 2]
```diff
@@ -139,7 +196,11 @@ def update_responses(self, updated_responses):
         If you do not want any writes to persist if called with unsupported keys,
         make sure to call in an atomic context.
         '''
-        # TODO: Add check for state once that stuff is here
+        if self.state is not ApprovalStates.IN_PROGRESS:
+            raise SchemaResponseStateError(
+                f'SchemaResponse has state f{self.reviews_state}. '
```
Comment (jwalz) [inline osf/models/schema_response.py:201]: "f{self.reviews_state}"?

File: osf/models/schema_response.py [hunk 3]
```diff
@@ -192,3 +264,158 @@ def _update_response(self, current_block, latest_response):
...
+    def _validate_trigger(self, event_data):
+        '''Any additional validation to confirm that a trigger is being used correctly.
+
+        For SchemaResponses, use this to confirm that the provided user has permission to
+        execute the trigger, including enforcing correct usage of the internal "accept" shortcut.
+        '''
+        user = event_data.kwargs.get('user')
+        trigger = event_data.event.name
+
+        # The only valid case for not providing a user is the internal accept shortcut
```
Comment (mfraezz) [inline osf/models/schema_response.py:287]: Minor: is there any precedent for, or would it make sense to add any flags to `event_data` in cases like these? e.g. `'auto': True` or `'internal': True`.

**Review by mfraezz — APPROVED** [review]
LGTM ⭐

**Pull request merged**